engine/server/session.go

```diff
@@ -1265,6 +1265,19 @@ func (srv *Server) ServeModule(ctx context.Context, mod *core.Module) error {
 	client.stateMu.Lock()
 	defer client.stateMu.Unlock()
 
+	// don't add the same module twice
+	// This can happen with generated clients since all remote dependencies are added
+	// on each connection and this could happen multiple times.
+	depMod, exist := client.deps.LookupDep(mod.Name())
+	if exist {
+		// Error if there's a conflict between dependencies
+		if depMod.GetSource().AsString() != "" && mod.Source.Self.AsString() != depMod.GetSource().AsString() {
+			return fmt.Errorf("module %s already exists with different source %s", mod.Name(), depMod.GetSource().AsString())
+		}
```

Comment:
`AsString` doesn't include the `Pin`.

Reply:
> AsString doesn't include the Pin. 

I think it does, it include the commit reference so that includes the pin no?

Reply:
Where does it include the commit reference?

https://github.com/dagger/dagger/blob/84ee1197610edd94b8997323bc7c7e7f379e288a/core/modulesource.go#L216-L227

The `version` doesn't neccessarily have the commit info. It *is* included in the field called `Pin`.

https://github.com/dagger/dagger/blob/84ee1197610edd94b8997323bc7c7e7f379e288a/core/modulesource.go#L493-L498